lib/matobjtriangularmatrices.gi

```diff
@@ -0,0 +1,599 @@
+                        filter and IsMutable), obj );
+     return obj;
+ end );
+
+ InstallMethod( NewIdentityMatrix,
+   "for IsUpperTriangularMatrixRep, a ring, and an int",
+   [ IsUpperTriangularMatrixRep, IsRing, IsInt ],
+   function( filter, basedomain, dim )
+     local mat, one, i;
+     mat := NewZeroMatrix(filter, basedomain, dim, dim);
+     one := One(basedomain);
+     for i in [1..dim] do
+         mat[i,i] := one;
+     od;
+     return mat;
+  end );
+
+ ############################################################################
+ # The basic attributes:
+ ############################################################################
+
+
+# BaseDomain, NrRows, NrCols can be directly obatined from the object
+ InstallMethod( BaseDomain, "for a IsUpperTriangularMatrixRep matrix",
+   [ IsUpperTriangularMatrixRep ],
+   function( m )
+     return m![UPPERTRIANGULARMATREP_BDPOS];
+end );
+
+ InstallMethod( NumberRows, "for a IsUpperTriangularMatrixRep matrix",
+   [ IsUpperTriangularMatrixRep ],
+   function( m )
+     return m![UPPERTRIANGULARMATREP_NRPOS];
+ end );
+
+ InstallMethod( NumberColumns, "for a IsUpperTriangularMatrixRep matrix",
+   [ IsUpperTriangularMatrixRep ],
+   function( m )
+     return m![UPPERTRIANGULARMATREP_NRPOS];
+  end );
+
+ InstallMethod( DimensionsMat, "for a IsUpperTriangularMatrixRep matrix",
+   [ IsUpperTriangularMatrixRep ],
+   function( m )
+     return [m![UPPERTRIANGULARMATREP_NRPOS],m![UPPERTRIANGULARMATREP_NRPOS]];
+end );
+
+ ############################################################################
+ # Representation preserving constructors:
+ ############################################################################
+
+ # Why are there extra implementations for these in matobjplist.gi?
+
+ ############################################################################
+ # A selection of list operations:
+ ############################################################################
+
+ # note depending on your particular matrix object implementing these might not
+ # make sense.
+
+ InstallOtherMethod( \[\], "for an IsUpperTriangularMatrixRep matrix and a positive integer",
+ #T Once the declaration of '\[\]' for 'IsMatrixObj' disappears,
+ #T we can use 'InstallMethod'.
+   [ IsUpperTriangularMatrixRep, IsPosInt ],
+   function( mat, row )
+   # could this cause problems if some entries in the vector are not bound?
+     local index_start, index_end, vec, i;
+     index_start := (row-1)*mat![UPPERTRIANGULARMATREP_NRPOS] + 1;
+     index_end := index_start + mat![UPPERTRIANGULARMATREP_NRPOS];
+     vec := NewZeroVector(IsPlistVectorRep,mat![UPPERTRIANGULARMATREP_BDPOS],mat![UPPERTRIANGULARMATREP_NRPOS]);
+     for i in [row..mat![UPPERTRIANGULARMATREP_NRPOS]] do 
+        vec[i] := mat![UPPERTRIANGULARMATREP_ELSPOS][(-row*row+row)/2+mat![UPPERTRIANGULARMATREP_NRPOS]*(row-1) + i];
+     od;
+     return NewVector(IsPlistVectorRep,mat![UPPERTRIANGULARMATREP_BDPOS],List(vec));
```

Comment:
`mat[i][j] := 5;` is not the same as `mat[i,j]:=5` now -- the former does not modify `mat`, the latter does... but historic code expects both to be the same...

My suggestion: return an *immutable* vector here, then the user at least gets an error.

(More generally: maybe we should not provide this at all, to prevent users from writing in-efficient code, *or* a generic function for this should be provided... dunnno)